flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java

```diff
@@ -383,18 +439,69 @@ public String toString() {
         }
     }
 
-    private static class SqlErrorTestItem extends ErrorTestItem {
-        final String expression;
+    private static class SqlErrorTestItem extends ErrorTestItem<String> {
 
         private SqlErrorTestItem(
                 String expression, String errorMessage, boolean expectedDuringValidation) {
-            super(errorMessage, expectedDuringValidation);
-            this.expression = expression;
+            super(expression, errorMessage, expectedDuringValidation);
+        }
+
+        @Override
+        String expression() {
+            return expression;
         }
 
         @Override
         public String toString() {
             return "[SQL] " + expression;
         }
     }
+
+    private static List<DataType> createDataTypes(
+            DataTypeFactory dataTypeFactory, AbstractDataType<?>[] dataTypes) {
+        return Stream.of(dataTypes)
+                .map(dataTypeFactory::createDataType)
+                .collect(Collectors.toList());
+    }
+
+    /** Helper POJO to store test parameters. */
+    public static class TableTestSpecColumn {
+
+        Expression tableApiExpression;
+        String sqlExpression;
+        Object result;
+        AbstractDataType<?> tableApiDataType;
+        AbstractDataType<?> sqlDataType;
+
+        public static TableTestSpecColumn of(
+                Expression tableApiExpression,
+                String sqlExpression,
+                Object result,
+                AbstractDataType<?> dataType) {
+            return of(tableApiExpression, sqlExpression, result, dataType, dataType);
+        }
+
+        public static TableTestSpecColumn of(
+                Expression tableApiExpression,
+                String sqlExpression,
+                Object result,
+                AbstractDataType<?> tableApiDataType,
+                AbstractDataType<?> sqlQueryDataType) {
```

Comment:
if Table API and SQL API share the same result it should in most cases also have the same data type?

Reply:
I kept this variant exposed as there are a couple of tests in `JsonFunctionsITCase` that make use of different return data types: https://github.com/apache/flink/pull/17341/files#diff-14f6e87fed00e28afef4a5e995126c430b4065661e3dc2f66d14f3741378f874R180

Reply:
TBH this is a very special case and hopefully does not happen elsewhere. we should not start having more of these special cases.